src/EdgeConProblem/EdgeConReduction.c

```diff
@@ -8,36 +8,290 @@
+    }
+
     return mk_bool_var(ctx, name);
 }
 
-Z3_ast getVariableParent(Z3_context ctx, int child, int parent)
-{
+Z3_ast getVariableParent(Z3_context ctx, int child, int parent) {
     char name[40];
+
     snprintf(name, 40, "p_[%d,%d]", child, parent);
+
     return mk_bool_var(ctx, name);
 }
 
-Z3_ast getVariableLevelInSpanningTree(Z3_context ctx, int level, int component)
-{
+Z3_ast getVariableLevelInSpanningTree(Z3_context ctx, int level, int component) {
     char name[40];
+
     snprintf(name, 40, "l_[%d,%d]", component, level);
+
     return mk_bool_var(ctx, name);
 }
 
-Z3_ast EdgeConReduction(Z3_context ctx, EdgeConGraph edgeGraph, int cost)
-{
-    return Z3_mk_false(ctx);
+Z3_ast EdgeConReduction(Z3_context z3_ctx, EdgeConGraph edgeGraph, int cost) {
+    g_context_s *ctx;
+
+    ctx = init_g_context(z3_ctx, edgeGraph, cost);
+
+    return(
+        OR(5)
+            build_phi_2(ctx),
+            build_phi_3(ctx),
+            build_phi_4(ctx),
+            build_phi_5(ctx),
+            build_phi_8(ctx)
+        EOR
+    );
+}
+
+static g_context_s* init_g_context(Z3_context z3_ctx, EdgeConGraph graph, int cost) {
+    g_context_s *ctx = NULL;
+
+    ctx = malloc(sizeof(g_context_s));
+    assert( NULL != ctx );
+
+    ctx->graph = graph;
+    ctx->G = getGraph(graph);
+    ctx->n = ctx->G.numNodes;
+    ctx->m = ctx->G.numEdges;
+    ctx->C_H = getNumComponents(graph);
+    ctx->N = ctx->C_H - 1;
+    ctx->k = cost;
+    ctx->z3_ctx = z3_ctx;
+
+    return ctx;
+}
+
+static Z3_ast build_phi_2(const g_context_s *ctx) { return Z3_mk_false(ctx->z3_ctx); }
+static Z3_ast build_phi_3(const g_context_s *ctx) { return Z3_mk_false(ctx->z3_ctx); }
+static Z3_ast build_phi_4(const g_context_s *ctx) { return Z3_mk_false(ctx->z3_ctx); }
+
+static Z3_ast build_phi_5(const g_context_s *ctx) {
+    int pos;
+    Z3_ast literals[
+        ctx->C_H *
+        (ctx->N <= ctx->k ? ctx->N : (ctx->N - ctx->k))
+    ];
+
+    pos = 0;
+    for (int i = 0; i < ctx->C_H; ++i) {
+        for (int n = ctx->k; n < ctx->N; ++n) {
+            literals[pos++] = L_(n, i);
+        }
+    }
+
+    return Z3_mk_or(ctx->z3_ctx, pos, literals);
+}
+
+static Z3_ast build_phi_6(const g_context_s *ctx, const int j1, const int j2) {
+    int pos;
+    Z3_ast literals[ctx->m * ctx->N];
+
+    pos = 0;
+    for (int u = 0; u < ctx->n; ++u) {
+        for (int v = u + 1; v < ctx->n; ++v) {
+            if (isEdge(ctx->G, u, v) &&
+                isNodeInComponent(ctx->graph, u, j1) &&
+                isNodeInComponent(ctx->graph, v, j2)) {
+                for (int i = 0; i < ctx->N; ++i) {
+                    literals[pos++] = X_(u, v, i);
+                }
+            }
+        }
+    }
+
+    return Z3_mk_or(ctx->z3_ctx, pos, literals);
+}
+
+static Z3_ast build_phi_7(const g_context_s *ctx, const int j1, const int j2) {
+    int pos;
+    Z3_ast literals[ctx->N];
+
+    pos = 0;
+    for (int h = 1; h < ctx->N; ++h) {
```

Comment:
h ne drevrait-il pas être initialisé à `0` ?
```suggestion
    for (int h = 0; h < ctx->N; ++h) {
```

Reply:
As discussed offline, this is an error in the second assignment.
I added the issue in the function comment:

https://github.com/EmileRolley/coca-project/blob/84d225a93b9ad9016b6dc908cc2e221bfb6beb73/src/EdgeConProblem/EdgeConReduction.c#L66-L73